### PR TITLE
gh-109740: Use 't' in `--disable-gil` SOABI

### PR DESCRIPTION
**Pull request by colesbury**
Shared libraries for CPython 3.13 `--disable-gil` builds are now marked with a 't' for threading. For example, `binascii.cpython-313t-darwin.so`.

* Issue: gh-109740

### CODE REVIEW

**Comment by colesbury** [comment]
Notes for reviewers:
* I moved the `--disable-gil` flag up in `configure` so that it comes before the debug check so that the optional "d" ABI flag appears on the outside (e.g., `package-cpython-313td-darwin.so`). This is purely aesthetic and can be changed, i fyou prefer.
* Contrary to what I originally though, `pypa/packaging` will need to be updated to compute CPython tags for `--disable-gil` builds.

File: Misc/NEWS.d/next/Build/2023-09-26-16-00-50.gh-issue-109740.wboWdQ.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+The experimental ``--disable-gil`` flag now includes "t" (for "threaded") in
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Build/2023-09-26-16-00-50.gh-issue-109740.wboWdQ.rst]:
```suggestion
The experimental ``--disable-gil`` configure flag now includes "t" (for "threaded") in
```

**Review by erlend-aasland — APPROVED** [review]

File: Lib/test/test_sys.py [hunk 2]
```diff
@@ -1210,6 +1210,13 @@ def test_pystats(self):
         sys._stats_clear()
         sys._stats_dump()
 
+    @test.support.cpython_only
+    @unittest.skipUnless(hasattr(sys, 'abiflags'), 'need sys.abiflags')
```
Comment (vstinner) [inline Lib/test/test_sys.py:1214]: I don't understand this skip. When sys.abiflags is not available? Nothing says in the doc that it's optional: https://docs.python.org/dev/library/sys.html#sys.abiflags

File: configure.ac [hunk 3]
```diff
@@ -1489,6 +1489,22 @@ fi
...
+AC_ARG_ENABLE([gil],
+  [AS_HELP_STRING([--disable-gil], [enable experimental support for running without the GIL (default is no)])],
+  [AS_VAR_IF([enable_gil], [yes], [disable_gil=no], [disable_gil=yes])], [disable_gil=no]
+)
+AC_MSG_RESULT([$disable_gil])
+
+if test "$disable_gil" = "yes"
+then
+  AC_DEFINE([Py_NOGIL], [1],
+            [Define if you want to disable the GIL])
```
Comment (vstinner) [inline configure.ac:1504]: Can you please give your explanation for the letter "t" here?

> "t" (for "threaded").

File: Lib/test/test_sys.py (hunk 2 above)
Comment (colesbury) [reply Lib/test/test_sys.py:1214]: It's only available on POSIX systems (not Windows).

Comment (vstinner) [reply Lib/test/test_sys.py:1214]: Oh, I didn't know. Would you mind to add ``   .. availability:: Unix.`` to https://docs.python.org/dev/library/sys.html#sys.abiflags ?

File: Python/dynload_win.c [hunk 4]
```diff
@@ -15,10 +15,16 @@
 #define PYD_DEBUG_SUFFIX ""
 #endif
 
+#ifdef Py_NOGIL
+#define PYD_THREADING_TAG "t"
+#else
+#define PYD_THREADING_TAG ""
+#endif
```
Comment (vstinner) [inline Python/dynload_win.c]:
```suggestion
#ifdef Py_NOGIL
#  define PYD_THREADING_TAG "t"
#else
#  define PYD_THREADING_TAG ""
#endif
```

File: configure.ac [hunk 5]
```diff
@@ -1489,6 +1489,23 @@ fi
...
+  # "t" (for "threaded")
```
Comment (vstinner) [inline configure.ac]:
```suggestion
  # Add "t" for "threaded"
```

**Review by vstinner — APPROVED** [review]
LGTM, but i left a few minor suggestions.

**Review by corona10 — APPROVED** [review]
lgtm!

File: Lib/test/test_sys.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_sys.py:1214]: I took the freedom to including this change as part of your PR.

**Comment by vstinner** [comment]
I included my suggested changes to be able to merge your PR more quickly ;-)

**Comment by vstinner** [comment]
> You need to regen configure.

Oops. Otherwise, it would be too easy! Done.

**Comment by colesbury** [comment]
Thanks!

**Comment by vstinner** [comment]
Windows x64 unstable tests:

> FAIL: test_future_times_out (test.test_concurrent_futures.test_as_completed.ThreadPoolAsCompletedTest.test_future_times_out) [0.01]

Should be already fixed by PR #109949.

> FAIL: test_set_and_clear (test.test_threading.EventTests.test_set_and_clear)

I created issue #109974 for this failure.

**Comment by vstinner** [comment]
The test broke test_importlib on Windows: see issue gh-110359.

```
FAIL: test_tagged_suffix (test.test_importlib.test_windows.Source_WindowsExtensionSuffixTests.test_tagged_suffix)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Administrator\buildarea\3.x.itamaro-win64-srv-22-aws.x64.nogil\build\Lib\test\test_importlib\test_windows.py", line 122, in test_tagged_suffix
    self.assertIn(expected_tag, suffixes)
AssertionError: '_d.cp313-win32.pyd' not found in ['_d.cp313t-win_amd64.pyd', '_d.pyd']
```

**Pull request merged**